### PR TITLE
new lookup plugin: LMDB

### PR DESCRIPTION
**Pull request by jpmens**
##### SUMMARY
This implements a lookup plugin for [LMDB](https://en.wikipedia.org/wiki/Lightning_Memory-Mapped_Database) databases using a required `pip install lmdb`. I have called it `lmdb_kv` because it cannot be named `lmdb` due to our having to import `lmdb` and the name goes well (I think) with, say, `redis_kv`.

This module is able to get specific keys and also enumerate the whole content of a database.

This PR also adds an example to the Ansible lookups Web site documentation.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lookup/lmdb_kv

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /Users/jpm/.ansible.cfg
  configured module search path = [u'/etc/ansible/library', u'/Users/jpm/Auto/projects/on-github/ansible/pdns_zone/library']
  python version = 2.7.13 (default, Apr 10 2017, 13:51:30) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]

```

#### create database

```python
#!/usr/bin/env python

import lmdb

map_size = 1024 * 100

env = lmdb.open('jp.mdb', map_size=map_size)

with env.begin(write=True) as txn:
    txn.put('fr', 'France')
    txn.put('nl', 'Netherlands')
    txn.put('es', 'Spain')
    txn.put('lu', 'Luxemburg')
```

#### playbook

```yaml
- hosts: localhost
  connection: local
  gather_facts: False
  vars:
    - benelux: "{{ lookup('lmdb_kv', 'nl', 'be', 'lu', db='jp.mdb') }}"
    - countries: "{{ lookup('lmdb_kv', db='jp.mdb') }}"
  tasks:
  - name: Get specific keys (in terms)
    debug: var=benelux

  - name: Nice country list
    debug: msg="Country code {{ item.0 | upper}} is actually {{ item.1 }}"
    with_items:
      - "{{ countries }}"

  - name: Template Europe
    template: src=europe.in dest=./output.txt

  - debug: msg="Hello from {{ item.0 }} a.k.a. {{ item.1 }}"
    with_lmdb_kv:
      - "n*"

```

#### template (europe.in)

```django
{% for cc,country in countries %}
  {{ cc|upper }} is actually {{ country }}
{% endfor %}
```

#### playbook run

```

PLAY [localhost] ***********************************************************************************

TASK [Get specific keys (in terms)] ****************************************************************
ok: [localhost] => {
    "benelux": "Netherlands,Luxemburg", 
    "changed": false
}

TASK [Nice country list] ***************************************************************************
ok: [localhost] => (item=('es', 'Spain')) => {
    "item": [
        "es", 
        "Spain"
    ], 
    "msg": "Country code ES is actually Spain"
}
ok: [localhost] => (item=('fr', 'France')) => {
    "item": [
        "fr", 
        "France"
    ], 
    "msg": "Country code FR is actually France"
}
ok: [localhost] => (item=('lu', 'Luxemburg')) => {
    "item": [
        "lu", 
        "Luxemburg"
    ], 
    "msg": "Country code LU is actually Luxemburg"
}
ok: [localhost] => (item=('nl', 'Netherlands')) => {
    "item": [
        "nl", 
        "Netherlands"
    ], 
    "msg": "Country code NL is actually Netherlands"
}

TASK [Template Europe] *****************************************************************************
ok: [localhost]

TASK [debug] ***************************************************************************************
ok: [localhost] => (item=('nl', 'Netherlands')) => {
    "item": [
        "nl", 
        "Netherlands"
    ], 
    "msg": "Hello from nl a.k.a. Netherlands"
}

PLAY RECAP *****************************************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   

```

#### template output

```
  ES is actually Spain
  FR is actually France
  LU is actually Luxemburg
  NL is actually Netherlands
```

### CODE REVIEW

**Comment by dharmabumstead** [comment]
Closing this one. Feel free to rebase and reopen.

**Comment by abadger** [comment]
Re-opening as this is a new lookup plugin, no need for closing it.

File: lib/ansible/plugins/lookup/lmdb_kv.py
Comment (bcoca) [inline lib/ansible/plugins/lookup/lmdb_kv.py:52]: We now require that lookup plugins have documentation to be included in Ansible

**Review by bcoca — DISMISSED** [review]

File: lib/ansible/plugins/lookup/lmdb_kv.py
Comment (jpmens) [reply lib/ansible/plugins/lookup/lmdb_kv.py:52]: a year later? honestly?

Comment (bcoca) [reply lib/ansible/plugins/lookup/lmdb_kv.py:52]: sorry for the delay, but this one was dropped in my queue as it was closed/reopppened

Comment (gundalow) [reply lib/ansible/plugins/lookup/lmdb_kv.py:52]: We now have docs team reviewing all the docs. 
@jpmens Do you have time to work on this?

Comment (jpmens) [reply lib/ansible/plugins/lookup/lmdb_kv.py:52]: I do not, @gundalow, sorry.

**Comment by gundalow** [comment]
If anyone is interested in this the PR needs updating to include the docs in the lookup plugin

**Comment by jpmens** [comment]
@gundalow I have added the documentation.

File: lib/ansible/plugins/lookup/lmdb_kv.py
Comment (acozine) [reply lib/ansible/plugins/lookup/lmdb_kv.py:52]: documentation added - thanks @jpmens

**Comment by acozine** [comment]
@jpmens this PR's title still says it's a work-in-progress, are you planning to change or expand it? Or is it ready for final review?

**Comment by jpmens** [comment]
@acozine it was, I suppose, a WIP one and a half years ago. I have added the documentation (a month ago) and do not intend working on this further.

File: lib/ansible/plugins/lookup/lmdb_kv.py
Comment (acozine) [inline lib/ansible/plugins/lookup/lmdb_kv.py:12]: @gundalow this would be a good PR for the community review session - if there's interest, the `version_added` would need to be updated.

**Comment by gundalow** [comment]
Who's interested in testing this PR?

**Comment by gundalow** [comment]
So we'd prefer not to take in a new lookup plugin without a maintainer or tests, so before this can be merged we are looking for:
* a maintainer
* some tests

File: lib/ansible/plugins/lookup/lmdb_kv.py
Comment (dw) [inline lib/ansible/plugins/lookup/lmdb_kv.py]: Despite the name (comes from the C API unfortunately), set_range() only locates the key or the next greatest key, it does not otherwise constrain the results produced by the cursor :) It needs to be something like:

````python
prefix = term[:-1] # strip asterisk
cursor.set_range(prefix)
while cursor.key().startswith(prefix):
    ret.append(cursor.item())
    cursor.next()
````

**Review by dw — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/lookup/lmdb_kv.py
Comment (jpmens) [reply lib/ansible/plugins/lookup/lmdb_kv.py]: Thank you David.

**Comment by gundalow** [comment]
I've pushed a commit to use `version_added: "2.8"`
rebuild_merge

**Comment by gundalow** [comment]
rebuild_merge

**Comment by resmo** [comment]
ready_for_review

**Comment by resmo** [comment]
added the missing things in #62670

**Pull request closed without merging**
